### DISCUSSION TITLE
Changing background color of existing PDF page

### DISCUSSION THREAD

**Discussion by usmansc** [pymupdf/PyMuPDF#2053] [Q&A] [answered]
I have a pdf from external source. Sometimes it has background color, sometimes it does not. I want it to be without the background color. I am trying to do this manually for ages, today I found pymupdf and I am wondering if such a thing is possible with it. I am not sure how it should be done with pymu, but manually I was trying to find the object that represented the background, delete the object and update xref table. Are such things possible with pymu?

**Comment by JorjMcKie** [comment, marked as answer]
> I have a pdf from external source. Sometimes it has background color, sometimes it does not. I want it to be without the background color. I am trying to do this manually for ages, today I found pymupdf and I am wondering if such a thing is possible with it. I am not sure how it should be done with pymu, but manually I was trying to find the object that represented the background, delete the object and update xref table. Are such things possible with pymu?

Yes, this is possible.
As always with PDF, there are amazingly many ways to achieve the type of thing you describe (incomplete list):
* Image used covering the page. Solution: find the image xref and replace with a small, empty, transparent image.
* PDF "line art" knows ways to generate filled rectangles. This may be very hard to detect - sometimes even impossible with any common sense effort. **_If_** these instructions can be detected however, then PyMuPDF has ways to delete them.
* PDF Watermark Artifacts are also part of the PDF spec. Easier to detect and remove than the previous case.

As you mentioned xref, your cases seem to belong to the first point above, though.
Please add more detail to your question.

**Reply by usmansc** [reply]
[GCC.pdf](https://github.com/pymupdf/PyMuPDF/files/10030009/GCC.pdf)
Let's say I have pdf like this and I want to remove green color. I want it to be transparent. This would most likely fit the first use case. Some other pdfs I have encountered had watermark, so the 3rd would be correct I assume.

**Reply by JorjMcKie** [reply]
Correct. Here is a snippet which removes the green stuff:

```python
import fitz
doc=fitz.open("GCC.pdf")
page=doc[0]  # load the page of interest
page.rect  # look at page dimensions
Rect(0.0, 0.0, 595.280029296875, 841.8900146484375)

print(page.get_images())  # check page's images
[(5, 0, 900, 600, 8, 'ICCBased', '', 'Im1', 'FlateDecode')]

page.get_image_rects(5)  # there is just one (xref 5), bbox is fairly large compared to page rect:
[Rect(72.6400146484375, 83.44500732421875, 522.6400146484375, 758.4450073242188)]

from replacer import img_replace  # import a utility function
# the following instructions create a tiny empty transparent pixmap
pix = fitz.Pixmap(fitz.csGRAY, (0, 0, 1, 1), 1)
pix.clear_with()

# replace image at xref 5 with the pixmap:
img_replace(page, 5, pixmap=pix)
doc.save("GCC-clear.pdf")
```

The script containing utility function `img_replace` can be found [here](https://github.com/pymupdf/PyMuPDF-Utilities/blob/master/image-replacement/replacer.py).
BTW we are considering to make this utility function official in a future version.

**Reply by JorjMcKie** [reply]
If you can provide an Artifact case, I hope I can show how this case works.

**Reply by JorjMcKie** [reply]
Forgot to mention:
To also **physically** get rid of the green image, use a save option that recovers its space: `doc.ez_save()` or `doc.save(filename, garbage=3,...)`.

**Reply by JorjMcKie** [reply]
It actually has no watermark, but drawings that look like text - but are not (that diagonal transparent name).
To delete this type of thing is practicaly impossible.

**Reply by nissansz** [reply]
How can I draw an image or white rectangular on a specified part of page?

**Reply by JorjMcKie** [reply]
Please do read the documentation!

* `page.draw_rect()` draws rectangle, there exist many more similar to draw other objects.
* `page.insert_image()` insert an image
* etc.

**Reply by nuthinking** [reply]
@JorjMcKie once all images are removed, is it possible to export a transparent PNG? I tried with `page.get_pixmap(matrix=mat, alpha=True)` and I still get the white background. I removed all `xrefs` and images are not visible anymore. I would like to export all text with a transparent background, basically. Thanks!

**Comment by JorjMcKie** [comment]
> @JorjMcKie once all images are removed, is it possible to export a transparent PNG? I tried with `page.get_pixmap(matrix=mat, alpha=True)` and I still get the white background. I removed all `xrefs` and images are not visible anymore. I would like to export all text with a transparent background, basically. Thanks!

You can make such an unremovable background yourself, e.g. `page.draw_rect(page.rect, color=None, fill=fitz.pdfcolor["yellow"], overlay=False)`.
If you are in a hacky mood, you may want to look at the page's `/Contents` object(s). At its start, some instruction doing this should be present ... which you can remove by editing and updating that object ...

**Reply by nuthinking** [reply]
> `page.draw_rect(page.rect, color=None, fill=fitz.pdfcolor["yellow"], overlay=False)`

This still leaves the white background when the PNG is exported. The `/Contents` object is as follows:
```
7 0 obj
<</Type/Page/Resources 11 0 R/MediaBox[0 7.8299815 595.5 850.07998]/Contents[12 0 R 169 0 R 170 0 R 171 0 R 172 0 R 173 0 R 174 0 R 175 0 R 176 0 R 177 0 R]/StructParents 0/Parent 2 0 R/Tabs/S/BleedBox[0 7.8299815 595.5 850.07998]/TrimBox[0 7.8299815 595.5 850.07998]/CropBox[0 7.8299815 595.5 850.07998]/Rotate 0/Annots[]>>
endobj
```

Any link where I can learn how to decode this? :)

**Reply by JorjMcKie** [reply]
Ah! A misconception: I did not mean to change **_your_** white background to yellow 😂.
What I meant was that at the start of your (original) first `/Contents` object, you will probably find the instructions that make a white background. And these instructions should look similar to what my example code also generates. That was all.

Assuming that your original first `/Contents` object is in xref 169, you can extract that into a `bytes` object like this: `c169 = doc.xref_stream(169)`. Then you may find instructions similar (!) to these:
```
0 0 595 842 re  % defines a rectangle (full page this case)
h  % command "close subpath" - may be missing
1 1 1 rg  % set fill color to RGB white
f  % fill path (= apply fill color to pevious stuff)
```
If you replace the `"f"` with a space and update the object 169 with the result like this `doc.update_stream(169, result)` your problem should be gone.

**Reply by nuthinking** [reply]
It doesn't look like my `/Contents` object contains any fill instructions. Looks more like a mask, you can see it above. Even if I search for "rg", "1 1 1", "white", or "255 255 255" in the whole document I don't find much. Looks like Acrobat [has a setting](https://pdf.wondershare.com/how-to/create-pdf-with-transparent-background.html) for it. No idea how this would reflect in the pdf file. I guess I need to install Acrobat and play with it. Are you sure it's not a matter of export? I will test the HTML export now. Thanks!

**Reply by nuthinking** [reply]
I inspected the file with Acrobat and indeed the issue is with those crop items I have in `/Contents`. Will try to figure out how to change them 💪

**Reply by nuthinking** [reply]
Problem is that if I try to edit the .pdf source on the Mac, for instance with Sublime, the file corrupts 😞

**Reply by JorjMcKie** [reply]
You cannot simply "edit" any PDF directly. That destroys the inner structure in a catastrophic way. Instead you must extract what you want to change, edit this and then re-insert (update). This leaves MuPDF a change to get all those pointers updated.

**Reply by nuthinking** [reply]
It corrupts even when I edit->undo->save. Meaning not making any changes. So I thought was a problem with file encoding. What a nasty format! :D

**Reply by JorjMcKie** [reply]
Absolutely yes. You probably also have a nasty editor (Mac!).

**Reply by nuthinking** [reply]
What would be an "empty" `/Contents` content I could replace mine with?

**Reply by nuthinking** [reply]
Using TextEdit doesn't break it. Was Sublime.

**Comment by JorjMcKie** [comment]
> What would be an "empty" `/Contents` content I could replace mine with?

If xref is the xref number of a `/Contents` object, you can empty it by simply doing `doc.update_stream(xref, b"")`.

**Reply by nuthinking** [reply]
Looks like the `MediaBox` is useful. No idea what else to remove from here:

```
<</Type/Page/Resources 11 0 R/MediaBox[0 7.8299815 595.5 850.07998]/Contents[12 0 R 179 0 R 180 0 R 181 0 R 182 0 R 183 0 R 184 0 R 185 0 R 186 0 R 187 0 R]/StructParents 0/Parent 2 0 R/Tabs/S/Rotate 0/Annots[]>>
```

I already removed the various `BleedBox`, `TrimBox` and `CropBox` 🤷‍♂️

**Reply by JorjMcKie** [reply]
You  mean in order to remove that background still?
Those poor boxes have nothing to do with it!
The code that causes the white background is very probably contained in xref 12 (see "12 0 R" above).

**Comment by JorjMcKie** [comment]
Why don't let me have that PDF so I can take a look a it ...

**Reply by nuthinking** [reply]
Check your email 🙏
